In [2]:
import pandas as pd

df = pd.read_csv('data.csv')

In [3]:
df.head(5)

Timestamp  playstation5  เครื่องทำขนมปัง  Ergonomic Wrist Rest  \
0  10/17/2021 17:29           NaN              NaN                   NaN   
1  10/17/2021 17:43           0.0              0.0                   0.0   
2  10/17/2021 17:43           0.0              0.0                   0.0   
3  10/17/2021 17:43           0.0              0.0                   0.0   
4  10/17/2021 17:43           0.0              0.0                   0.0   

   เครื่องอบผ้า  เครื่องชงกาแฟแคปซูล  เก้าอี้ LA-Z-Boy  \
0           NaN                  NaN               NaN   
1           0.0                  0.0               0.0   
2           0.0                  0.0               0.0   
3           1.0                  1.0               0.0   
4           0.0                  0.0               0.0   

   เครื่องให้อาหารสัตว์อัตโนมัติ  บัตตาเลี่ยน  แก้วเก็บความเย็น  ...  \
0                            NaN          NaN               NaN  ...   
1                            0.0          1.0               1.0  ...   
2                            0.0          0.0               1.0  ...   
3                            0.0          1.0               1.0  ...   
4                            0.0          0.0               1.0  ...   

   การ์ดจอ RTX 3080  ขนมจีนน้ำยาปู  Salmon Sashimi  จักรยานเสือหมอบ  \
0               NaN            NaN             NaN              NaN   
1               1.0            0.0             0.0              0.0   
2               0.0            0.0             1.0              0.0   
3               0.0            1.0             1.0              0.0   
4               0.0            0.0             1.0              0.0   

   ไฟแต่งห้องมินิมอล  External Harddisk  หม้อทอดไร้น้ํามัน  airpods   ยาดม  \
0                NaN                NaN                NaN       NaN   NaN   
1                0.0                1.0                1.0       0.0   0.0   
2                0.0                1.0                1.0       0.0   0.0   
3                1.0                1.0                1.0       1.0   1.0   
4                0.0                1.0                1.0       0.0   0.0   

   ไฟส่องหน้าไลฟ์สด  
0               NaN  
1               0.0  
2               0.0  
3               1.0  
4               0.0  

[5 rows x 42 columns]

In [4]:
df = df.drop(['Timestamp', 'playstation5'], axis=1)
df.head(5)

เครื่องทำขนมปัง  Ergonomic Wrist Rest  เครื่องอบผ้า  เครื่องชงกาแฟแคปซูล  \
0              NaN                   NaN           NaN                  NaN   
1              0.0                   0.0           0.0                  0.0   
2              0.0                   0.0           0.0                  0.0   
3              0.0                   0.0           1.0                  1.0   
4              0.0                   0.0           0.0                  0.0   

   เก้าอี้ LA-Z-Boy  เครื่องให้อาหารสัตว์อัตโนมัติ  บัตตาเลี่ยน  \
0               NaN                            NaN          NaN   
1               0.0                            0.0          1.0   
2               0.0                            0.0          0.0   
3               0.0                            0.0          1.0   
4               0.0                            0.0          0.0   

   แก้วเก็บความเย็น  ลู่วิ่งออกกำลังกาย  Kindle  ...  การ์ดจอ RTX 3080  \
0               NaN                 NaN     NaN  ...               NaN   
1               1.0                 0.0     0.0  ...               1.0   
2               1.0                 0.0     0.0  ...               0.0   
3               1.0                 1.0     0.0  ...               0.0   
4               1.0                 0.0     0.0  ...               0.0   

   ขนมจีนน้ำยาปู  Salmon Sashimi  จักรยานเสือหมอบ  ไฟแต่งห้องมินิมอล  \
0            NaN             NaN              NaN                NaN   
1            0.0             0.0              0.0                0.0   
2            0.0             1.0              0.0                0.0   
3            1.0             1.0              0.0                1.0   
4            0.0             1.0              0.0                0.0   

   External Harddisk  หม้อทอดไร้น้ํามัน  airpods   ยาดม  ไฟส่องหน้าไลฟ์สด  
0                NaN                NaN       NaN   NaN               NaN  
1                1.0                1.0       0.0   0.0               0.0  
2                1.0                1.0       0.0   0.0               0.0  
3                1.0                1.0       1.0   1.0               1.0  
4                1.0                1.0       0.0   0.0               0.0  

[5 rows x 40 columns]

In [5]:
df = df.drop([0])
df.head(5)

เครื่องทำขนมปัง  Ergonomic Wrist Rest  เครื่องอบผ้า  เครื่องชงกาแฟแคปซูล  \
1              0.0                   0.0           0.0                  0.0   
2              0.0                   0.0           0.0                  0.0   
3              0.0                   0.0           1.0                  1.0   
4              0.0                   0.0           0.0                  0.0   
5              0.0                   0.0           1.0                  0.0   

   เก้าอี้ LA-Z-Boy  เครื่องให้อาหารสัตว์อัตโนมัติ  บัตตาเลี่ยน  \
1               0.0                            0.0          1.0   
2               0.0                            0.0          0.0   
3               0.0                            0.0          1.0   
4               0.0                            0.0          0.0   
5               1.0                            0.0          1.0   

   แก้วเก็บความเย็น  ลู่วิ่งออกกำลังกาย  Kindle  ...  การ์ดจอ RTX 3080  \
1               1.0                 0.0     0.0  ...               1.0   
2               1.0                 0.0     0.0  ...               0.0   
3               1.0                 1.0     0.0  ...               0.0   
4               1.0                 0.0     0.0  ...               0.0   
5               1.0                 0.0     0.0  ...               0.0   

   ขนมจีนน้ำยาปู  Salmon Sashimi  จักรยานเสือหมอบ  ไฟแต่งห้องมินิมอล  \
1            0.0             0.0              0.0                0.0   
2            0.0             1.0              0.0                0.0   
3            1.0             1.0              0.0                1.0   
4            0.0             1.0              0.0                0.0   
5            0.0             1.0              1.0                0.0   

   External Harddisk  หม้อทอดไร้น้ํามัน  airpods   ยาดม  ไฟส่องหน้าไลฟ์สด  
1                1.0                1.0       0.0   0.0               0.0  
2                1.0                1.0       0.0   0.0               0.0  
3                1.0                1.0       1.0   1.0               1.0  
4                1.0                1.0       0.0   0.0               0.0  
5                1.0                1.0       0.0   1.0               0.0  

[5 rows x 40 columns]

In [6]:
missing = df.isnull().sum()
print(missing)

เครื่องทำขนมปัง                  0
Ergonomic Wrist Rest             0
เครื่องอบผ้า                     0
เครื่องชงกาแฟแคปซูล              0
เก้าอี้ LA-Z-Boy                 0
เครื่องให้อาหารสัตว์อัตโนมัติ    0
บัตตาเลี่ยน                      1
แก้วเก็บความเย็น                 0
ลู่วิ่งออกกำลังกาย               0
Kindle                           0
เครื่องซักผ้า                    0
Bluetooth Speaker                0
ห้องน้ำแมวอัตโนมัติ              0
PS5                              0
ทรายแมว                          0
ลำโพง pixel                      0
Logitech Mx Master 3 Mouse       0
ตุ๊กตา ty                        0
น้ำพุแมว                         0
Robot ดูดฝุ่น                    0
Mechanical keyboard              1
Nintendo switch                  1
หนังสือ python                   0
gaming chair                     1
Deskmat                          0
Dew - ไฟโรเซ่                    0
เทียนหอม jo malone               0
กระติกน้ำ 2 ลิตร                 1
ที่นอน memory form  

In [7]:
df = df.fillna(0)

In [8]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1
df = df.applymap(encode_units)

In [9]:
df.head(5)

เครื่องทำขนมปัง  Ergonomic Wrist Rest  เครื่องอบผ้า  เครื่องชงกาแฟแคปซูล  \
1                0                     0             0                    0   
2                0                     0             0                    0   
3                0                     0             1                    1   
4                0                     0             0                    0   
5                0                     0             1                    0   

   เก้าอี้ LA-Z-Boy  เครื่องให้อาหารสัตว์อัตโนมัติ  บัตตาเลี่ยน  \
1                 0                              0            1   
2                 0                              0            0   
3                 0                              0            1   
4                 0                              0            0   
5                 1                              0            1   

   แก้วเก็บความเย็น  ลู่วิ่งออกกำลังกาย  Kindle  ...  การ์ดจอ RTX 3080  \
1                 1                   0       0  ...                 1   
2                 1                   0       0  ...                 0   
3                 1                   1       0  ...                 0   
4                 1                   0       0  ...                 0   
5                 1                   0       0  ...                 0   

   ขนมจีนน้ำยาปู  Salmon Sashimi  จักรยานเสือหมอบ  ไฟแต่งห้องมินิมอล  \
1              0               0                0                  0   
2              0               1                0                  0   
3              1               1                0                  1   
4              0               1                0                  0   
5              0               1                1                  0   

   External Harddisk  หม้อทอดไร้น้ํามัน  airpods   ยาดม  ไฟส่องหน้าไลฟ์สด  
1                  1                  1         0     0                 0  
2                  1                  1         0     0                 0  
3                  1                  1         1     1                 1  
4                  1                  1         0     0                 0  
5                  1                  1         0     1                 0  

[5 rows x 40 columns]

In [35]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = apriori(df, min_support=0.1, use_colnames=True)

In [43]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules["antecedents_len"] = rules["antecedents"].apply(lambda x: len(x))
rules["consequents_len"] = rules["consequents"].apply(lambda x: len(x))
rules.head(10)

antecedents                   consequents  \
0                (เครื่องอบผ้า)            (แก้วเก็บความเย็น)   
1            (แก้วเก็บความเย็น)                (เครื่องอบผ้า)   
2                (เครื่องอบผ้า)               (เครื่องซักผ้า)   
3               (เครื่องซักผ้า)                (เครื่องอบผ้า)   
4                (เครื่องอบผ้า)           (Bluetooth Speaker)   
5           (Bluetooth Speaker)                (เครื่องอบผ้า)   
6            (กระติกน้ำ 2 ลิตร)                (เครื่องอบผ้า)   
7                (เครื่องอบผ้า)            (กระติกน้ำ 2 ลิตร)   
8                (เครื่องอบผ้า)  (พลาสเตอร์บรรเทาปวด ตราเสือ)   
9  (พลาสเตอร์บรรเทาปวด ตราเสือ)                (เครื่องอบผ้า)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.130435            0.717391  0.130435    1.000000  1.393939   
1            0.717391            0.130435  0.130435    0.181818  1.393939   
2            0.130435            0.521739  0.130435    1.000000  1.916667   
3            0.521739            0.130435  0.130435    0.250000  1.916667   
4            0.130435            0.739130  0.108696    0.833333  1.127451   
5            0.739130            0.130435  0.108696    0.147059  1.127451   
6            0.217391            0.130435  0.108696    0.500000  3.833333   
7            0.130435            0.217391  0.108696    0.833333  3.833333   
8            0.130435            0.565217  0.108696    0.833333  1.474359   
9            0.565217            0.130435  0.108696    0.192308  1.474359   

   leverage  conviction  antecedents_len  consequents_len  
0  0.036862         inf                1                1  
1  0.036862    1.062802                1                1  
2  0.062382         inf                1                1  
3  0.062382    1.159420                1                1  
4  0.012287    1.565217                1                1  
5  0.012287    1.019490                1                1  
6  0.080340    1.739130                1                1  
7  0.080340    4.695652                1                1  
8  0.034972    2.608696                1                1  
9  0.034972    1.076605                1                1

In [46]:
new_rules = rules[ (rules['lift'] >= 1) &
       (rules['confidence'] >= 0.5) &
       (rules['antecedents_len'] == 1) &
       (rules['consequents_len'] == 1) ]
new_rules.head(5)

antecedents          consequents  antecedent support  \
0      (เครื่องอบผ้า)   (แก้วเก็บความเย็น)            0.130435   
2      (เครื่องอบผ้า)      (เครื่องซักผ้า)            0.130435   
4      (เครื่องอบผ้า)  (Bluetooth Speaker)            0.130435   
6  (กระติกน้ำ 2 ลิตร)       (เครื่องอบผ้า)            0.217391   
7      (เครื่องอบผ้า)   (กระติกน้ำ 2 ลิตร)            0.130435   

   consequent support   support  confidence      lift  leverage  conviction  \
0            0.717391  0.130435    1.000000  1.393939  0.036862         inf   
2            0.521739  0.130435    1.000000  1.916667  0.062382         inf   
4            0.739130  0.108696    0.833333  1.127451  0.012287    1.565217   
6            0.130435  0.108696    0.500000  3.833333  0.080340    1.739130   
7            0.217391  0.108696    0.833333  3.833333  0.080340    4.695652   

   antecedents_len  consequents_len  
0                1                1  
2                1                1  
4                1                1  
6                1                1  
7                1                1

In [47]:
new_rules.sort_values(by=['lift'] ,ascending=False).head(20)

antecedents                   consequents  \
194                     (ทรายแมว)                    (น้ำพุแมว)   
195                    (น้ำพุแมว)                     (ทรายแมว)   
6              (กระติกน้ำ 2 ลิตร)                (เครื่องอบผ้า)   
7                  (เครื่องอบผ้า)            (กระติกน้ำ 2 ลิตร)   
19   (Logitech Mx Master 3 Mouse)         (เครื่องชงกาแฟแคปซูล)   
245             (Nintendo switch)              (หนังสือ python)   
230                (gaming chair)         (Mechanical keyboard)   
243            (ไฟส่องหน้าไลฟ์สด)         (Mechanical keyboard)   
114          (ลู่วิ่งออกกำลังกาย)                    (airpods )   
2                  (เครื่องอบผ้า)               (เครื่องซักผ้า)   
132  (Logitech Mx Master 3 Mouse)               (เครื่องซักผ้า)   
286          (เทียนหอม jo malone)           (หม้อทอดไร้น้ํามัน)   
303          (ที่นอน memory form)  (พลาสเตอร์บรรเทาปวด ตราเสือ)   
102          (ลู่วิ่งออกกำลังกาย)              (หนังสือ python)   
214            (กระติกน้ำ 2 ลิตร)               (Robot ดูดฝุ่น)   
146            (กระติกน้ำ 2 ลิตร)               (เครื่องซักผ้า)   
144          (เทียนหอม jo malone)               (เครื่องซักผ้า)   
138             (Nintendo switch)               (เครื่องซักผ้า)   
130                     (ทรายแมว)               (เครื่องซักผ้า)   
290            (กระติกน้ำ 2 ลิตร)  (พลาสเตอร์บรรเทาปวด ตราเสือ)   

     antecedent support  consequent support   support  confidence      lift  \
194            0.130435            0.108696  0.108696    0.833333  7.666667   
195            0.108696            0.130435  0.108696    1.000000  7.666667   
6              0.217391            0.130435  0.108696    0.500000  3.833333   
7              0.130435            0.217391  0.108696    0.833333  3.833333   
19             0.130435            0.347826  0.108696    0.833333  2.395833   
245            0.130435            0.500000  0.130435    1.000000  2.000000   
230            0.108696            0.500000  0.108696    1.000000  2.000000   
243            0.130435            0.500000  0.130435    1.000000  2.000000   
114            0.173913            0.456522  0.152174    0.875000  1.916667   
2              0.130435            0.521739  0.130435    1.000000  1.916667   
132            0.130435            0.521739  0.130435    1.000000  1.916667   
286            0.152174            0.543478  0.152174    1.000000  1.840000   
303            0.130435            0.565217  0.130435    1.000000  1.769231   
102            0.173913            0.500000  0.152174    0.875000  1.750000   
214            0.217391            0.347826  0.130435    0.600000  1.725000   
146            0.217391            0.521739  0.195652    0.900000  1.725000   
144            0.152174            0.521739  0.130435    0.857143  1.642857   
138            0.130435            0.521739  0.108696    0.833333  1.597222   
130            0.130435            0.521739  0.108696    0.833333  1.597222   
290            0.217391            0.565217  0.195652    0.900000  1.592308   

     leverage  conviction  antecedents_len  consequents_len  
194  0.094518    5.347826                1                1  
195  0.094518         inf                1                1  
6    0.080340    1.739130                1                1  
7    0.080340    4.695652                1                1  
19   0.063327    3.913043                1                1  
245  0.065217         inf                1                1  
230  0.054348         inf                1                1  
243  0.065217         inf                1                1  
114  0.072779    4.347826                1                1  
2    0.062382         inf                1                1  
132  0.062382         inf                1                1  
286  0.069471         inf                1                1  
303  0.056711         inf                1                1  
102  0.065217    4.000000                1                1  
214  0.054820    1.63043

สรุป
จากการ run association rules พบว่ามี item คู่ที่น่าสนใจได้แก่
rule 195 น้ำพุแมว แล้วจะมีการซื้อ ทรายแมว เป็นโปรโมชั่นไว้เสนอขายสำหรับคนรักแมวได้อาจจะเป็นโปรโมชั่นที่ว่า ซื้อน้ำพุแมวที่มีราคาสูง แล้วสามารถซื้อทรายแมวได้ในราคาถูกลงเป็นต้น
อีกคู่นึงที่น่าสนใจคือ rule 230 gaming chair และจะมีการซื้อ Mechanical keyboard น่าจะพออนุมาณได้ว่าเป็นกลุ่มคนชอบเล่นเกมส์ เราก็อาจจะจัดโปรโมชั่นเป็นต้นว่า ซื้อเก้าอี้แล้วได้ส่วนลดคีย์บอร์ดต่อได้